From 459dcfaa27b9b405e40a44501e14d4d24b795910 Mon Sep 17 00:00:00 2001
From: Anson Huang <b20788@freescale.com>
Date: Wed, 25 May 2011 19:03:19 +0800
Subject: [PATCH 581/670] ENGR38811709 [MX28]GPIO bank0 irq

i.MX28 GPIO bank 0 interrupts not functional in Linux BSP,
add work around, it may impact performance a little, but
acceptable.

Signed-off-by: Anson Huang <b20788@freescale.com>
---
 arch/arm/plat-mxs/include/mach/entry-macro.S | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/arch/arm/plat-mxs/include/mach/entry-macro.S b/arch/arm/plat-mxs/include/mach/entry-macro.S
index 353a7b2..2cc6458 100644
--- a/arch/arm/plat-mxs/include/mach/entry-macro.S
+++ b/arch/arm/plat-mxs/include/mach/entry-macro.S
@@ -1,7 +1,7 @@
 /*
  * Low-level IRQ helper macros for Freescale MXS-based
  *
- * Copyright (C) 2009-2010 Freescale Semiconductor, Inc. All Rights Reserved.
+ * Copyright (C) 2009-2011 Freescale Semiconductor, Inc. All Rights Reserved.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -26,11 +26,23 @@
 	ldr \base, [\base]
 	ldr \irqnr, [\base, #0x70]
 	cmp \irqnr, #0x7F
-	moveqs \irqnr, #0
+	bne 100f
+	/* If the irq is 0x7F, check the raw bit to see
+	whether it is a real GPIO0 irq */
+	ldr \tmp, [\base, #0xD0]
+	mov \tmp, \tmp, lsr #31
+	and \tmp, \tmp, #0x1
+	cmp \tmp, #0x1
+	bne 200f
+	/* Do a dummy compare to make sure Z = 0 */
+	cmp \tmp, #0xFF
+	b 100f
+200:	movs \irqnr, #0
+100:
 	.endm
 
 	.macro  get_irqnr_preamble, base, tmp
 	.endm
 
 	.macro  arch_ret_to_user, tmp1, tmp2
-	.endm
+.endm
-- 
1.8.0

